### PR TITLE
Add proxy and env file for Docker in bootstrap

### PR DESCRIPTION
**Pull request by nicklathe**
- Add proxy for Docker to use when pulling images during bootstrap
      step.
    - Cleanup after bootstrap script is finished

### CODE REVIEW

**Comment by nicklathe** [comment]
I'm testing this on an instance with Merit to verify the steps work to pull a Docker image

**Comment by nicklathe** [comment]
@dragan @weitenghuang: Here is a list of changes made:
    - Due to the docker.service in systemd not liking proxy variables set to empty strings (i.e. `HTTP_PROXY=""`), I had to templatize the entire docker proxy setting in the bootstrap scripts
    - Change the required variables to be `http_proxy`, `https_proxy`, and `no_proxy`, then use these variables in the numerous templates to format the correct way.
    - Move the above mentioned variables in /etcd/etcd.tf to /etcd/main.tf, since they are now referenced in both files.
    - Use the terraform ternary expression to assess whether to pass in a rendered template or just an empty string (for cases when there is no proxy needed).

File: etcd/environment/docker_proxy.config [hunk 1]
```diff
@@ -0,0 +1,3 @@
+HTTP_PROXY=${http_proxy}
+        HTTPS_PROXY=${https_proxy}
```
Comment (weitenghuang) [inline etcd/environment/docker_proxy.config]: is it required to indent starting from 2nd line?

File: etcd/bootstrapper/s3-cloudconfig-bootstrap.sh [hunk 2]
```diff
@@ -65,6 +71,8 @@ docker ps -aq | xargs -r docker rm
 docker volume ls -q | xargs -r docker volume rm
 docker images -q | xargs -r docker rmi
 systemctl stop docker.service
+rm -rf /etc/systemd/system/docker.service.d
+rm /etc/docker-environment.env
```
Comment (weitenghuang) [inline etcd/bootstrapper/s3-cloudconfig-bootstrap.sh]: if we already have `docker-environment.env` here, why will we need to re-create it at cloud-configs yml file?

File: etcd/bootstrapper/s3-cloudconfig-bootstrap.sh [hunk 3]
```diff
@@ -30,6 +30,12 @@ cloudinit_dst="$work_dir"
 cloudinit_common_tls_src="s3://$bucket/cloudinit/common/tls"
 cloudinit_tls_dst="$work_dir/tls"
 
+# Set proxy for Docker to use. If no proxy is supplied, this will be an empty string
+${docker_service_proxy}
+
+# Set environment variables for Docker to use. If no variables are supplied, file will be empty
+echo -e "${docker_environment}" > /etc/docker-environment.env
```
Comment (weitenghuang) [inline etcd/bootstrapper/s3-cloudconfig-bootstrap.sh]: shouldn't `${docker_environment}` be `${docker_proxy}`?

Comment (weitenghuang) [reply etcd/bootstrapper/s3-cloudconfig-bootstrap.sh]: I got why we have 2 different variables. Sorry, please ignore my comments. Thanks

File: etcd/bootstrapper/s3-cloudconfig-bootstrap.sh (hunk 2 above)
Comment (nicklathe) [reply etcd/bootstrapper/s3-cloudconfig-bootstrap.sh]: I'm just cleaning up after the bootstrap runs. I know cloud-init will rewrite this file, but I figured it we would let cloud-init do it's thing and write the new file.

File: etcd/environment/docker_proxy.config (hunk 1 above)
Comment (nicklathe) [reply etcd/environment/docker_proxy.config]: Yes, this is purposeful. This is due to the variable called in the cloud-configs being spaced 8 spaces, then the following lines here also need to be spaced 8 spaces. So this:
```
        HTTP_PROXY...
        HTTPS_PROXY...
        NO_PROXY...
```
was spacing the first line 8 spaces, then the following 2 lines 16 spaces (double).

File: etcd/bootstrapper/s3-cloudconfig-bootstrap.sh (hunk 2 above)
Comment (dragan) [reply etcd/bootstrapper/s3-cloudconfig-bootstrap.sh]: @weitenghuang Logic here should only support grabbing the cloud-config and then cleaning up after itself. The cloud-config then controls how the instance is configured.

File: etcd/environment/docker_proxy.config (hunk 1 above)
Comment (weitenghuang) [reply etcd/environment/docker_proxy.config]: I see now.
How about add an empty 1st line in this file so we can align the content?

File: etcd/bootstrapper/s3-cloudconfig-bootstrap.sh (hunk 2 above)
Comment (weitenghuang) [reply etcd/bootstrapper/s3-cloudconfig-bootstrap.sh]: make sense. thank you both for the explanation.

File: etcd/environment/docker_proxy.config (hunk 1 above)
Comment (nicklathe) [reply etcd/environment/docker_proxy.config]: I would be in support of that idea. @dragan?

Comment (dragan) [reply etcd/environment/docker_proxy.config]: I like it as well, but we need to test to make sure it works with the empty line.

**Comment by nicklathe** [comment]
@dragan @weitenghuang PR should be ready to go. We decided to leave the indentation of the template files as is for now, so as to let Receipts get back to deploying as quickly as possible. We can revisit later.

**Comment by nicklathe** [comment]
Forgot one last feature: a retry loop for Docker pull. We are experiencing _extremely_ slow connection during peak business hours, so we implemented a retry loop to handle timeouts.
